examples/file-sharing.rs

```diff
@@ -0,0 +1,639 @@
+                                    providers,
+                                    closest_peers,
+                                    ..
+                                })),
+                            ..
+                        },
+                    )) => {
+                        if let Some(sender) = outstanding_get_providers.remove(&id) {
+                            sender.send(providers).unwrap();
+                        }
+                    }
+                    SwarmEvent::Behaviour(ComposedEvent::Kademlia(_)) => {}
+                    SwarmEvent::Behaviour(ComposedEvent::RequestResponse(
+                        RequestResponseEvent::Message { peer, message },
+                    )) => match message {
+                        RequestResponseMessage::Request {
+                            request, channel, ..
+                        } => {
+                            event_sender
+                                .send(Event::InboundRequest {
+                                    request: request.0,
+                                    channel,
+                                })
+                                .await
+                                .unwrap();
+                        }
+                        RequestResponseMessage::Response {
+                            request_id,
+                            response,
+                        } => {
+                            outstanding_request_file
+                                .remove(&request_id)
+                                .expect("Request to still be oustanding.")
+                                .send(response.0)
+                                .unwrap();
+                        }
+                    },
+                    SwarmEvent::Behaviour(ComposedEvent::RequestResponse(
+                        RequestResponseEvent::ResponseSent { .. },
+                    )) => {}
+                    SwarmEvent::NewListenAddr { address, .. } => {
+                        let local_peer_id = *swarm.local_peer_id();
+                        println!(
+                            "Local node is listening on {:?}",
+                            address.with(Protocol::P2p(local_peer_id.into()))
+                        );
+                    }
+                    SwarmEvent::IncomingConnection { .. } => {}
+                    SwarmEvent::ConnectionEstablished {
+                        peer_id, endpoint, ..
+                    } => {
+                        if endpoint.is_dialer() {
+                            if let Some(sender) = outstanding_dial.remove(&peer_id) {
+                                sender.send(());
+                            }
+                        }
+                    }
+                    SwarmEvent::ConnectionClosed { .. } => {}
+                    SwarmEvent::UnreachableAddr { .. } => {}
+                    e => panic!("{:?}", e),
+                },
+                command = command_receiver.next() => match command {
```

Comment:
I am in favor of splitting the large `select!`, though moving it to separate functions would require passing the `pending_*` along with each call. With that in mind, I am not sure it would be an improvement.

Reply:
Or one could use a custom struct wrapping all state providing custom functions to interact with said state. This might also yield the opportunity to DRY.

Reply:
Thanks for the input @elenaf9 and @wngr. e05117d092a949bb5fa0e77e26a3c5994c91939f splits the `select!` into separate `async fn`s.